gadopt/approximations.py

```diff
@@ -412,15 +420,93 @@ def buoyancy(self, p, T):
+class BaseGIAApproximation:
+    """Base class for viscoelasticity assuming small displacements.
+
+    By assuming a small displacement with respect to mantle thickness
+    we can linearise the problem, introducing a perturbation away from a reference state.
+
+    For background and derivation of the formulation please see the equations and
+    references provided in Scott et al 2025.
+
+    Automated forward and adjoint modelling of viscoelastic deformation of the solid
+    Earth.  Scott, W.; Hoggard, M.; Duvernay, T.; Ghelichkhan, S.; Gibson, A.;
+    Roberts, D.; Kramer, S. C.; and Davies, D. R. EGUsphere, 2025: 1–43. 2025.
+
+    Arguments:
+      density:       density of the reference state - assumed to be hydrostatic
+      shear_modulus: shear modulus
+      viscosity:     viscosity
+      g:             gravitational acceleration
+      B_mu:          Nondimensional number describing ratio of buoyancy to elastic
+                     shear strength used for nondimensionalisation.
+                     $$ B_{\\mu} = \frac{\bar{\rho} \bar{g} L}{\bar{\\mu}}$,
+                     where $\bar{\rho}$ is a characteristic density scale (kg / m^3),
+                     $\bar{g}$ is a characteristic gravity scale (m / s^2),
+                     $L$ is a characteristic length scale, often Mantle depth (m),
+                     $\\mu$ is a characteristic shear modulus (Pa).
+    """
+
+    def __init__(
+        self,
+        density: Function | Number,
+        shear_modulus: Function | Number | list,
+        viscosity: Function | Number | list,
+        *,
+        g: Function | Number = 1,
+        B_mu: Function | Number = 1,
+    ):
+        self.density = ensure_constant(density)
+        self.shear_modulus = ensure_constant(shear_modulus)
+        self.viscosity = ensure_constant(viscosity)
+        self.g = ensure_constant(g)
+        self.B_mu = ensure_constant(B_mu)
+
+    def buoyancy(self, displacement):
+        # Buoyancy term due to the advection of the background density field
+        # written on RHS of equations
+        return self.B_mu * self.g * inner(displacement, grad(self.density))
 
-    By assuming a small displacement, we can linearise the problem, assuming a perturbation
-    away from a reference state. We assume a Maxwell viscoelastic rheology, i.e.
-    stress is the same but viscous and elastic strains combine linearly. We follow
-    the approach by Zhong et al. 2003 redefining the problem in terms of incremental
+    def rho_continuity(self):
+        # StokesSolverBase in stokes_integrators.py currently requires rho_continuity
+        # from approximation. This is only strictly necessary for quasi-compressible
+        # approximations in Mantle convection e.g. TALA and ALA, but is used in setting up
+        # the continuity equation for the IncompressibleMaxwellApproximation to use similar
+        # code for the incompressible viscous mantle convection in StokesSolverBase.
+        return 1
```

Comment:
`rho_continuity` has two purposes in `StokesSolverBase`: to define an element of the application context and to define an equation attribute used in `divergence_term`. For the latter, this is an easy fix: the `Equation` instance holds the approximation, so `rho_continuity` can be directly called in `divergence_term`. For the former, one way around it is to define an empty application context in `StokesSolverBase` and then have each solver add to it based on its requirements. Then we should be able to get rid of this method.